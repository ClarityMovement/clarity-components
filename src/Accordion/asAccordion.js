import React from 'react';
import _ from 'lodash';
import { compose, withState, withProps, withHandlers } from 'recompose';
import { embedHandler } from 'react-render-counter/hocs';
import asAccordionItem from './asAccordionItem';

const rootStyles = {
  default: {
    padding: 0,
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
    width: '100%',
  },
  expanded: {
    height: '100%',
  },
  wide: {
    width: '95%',
    margin: '0 auto',
  },
};

const stylesMapper = ({ wide, expanded, style }) => {
  const rootStyle = { ...rootStyles.default };
  if (wide) {
    Object.assign(rootStyle, rootStyles.wide);
  }
  if (expanded) {
    Object.assign(rootStyle, rootStyles.expanded);
  }

  return {
    style: { ...rootStyle, ...style },
  };
};

const propsMapper = ({ children, ...containerProps }) => {
  const toItem = (element, index) => {
    const Item = asAccordionItem(element.type);
    const defaultProps = element.type.defaultProps || {};
    const combinedProps = {
      ...containerProps,
      ..._.omitBy(element.props, (value, key) => value === defaultProps[key]),
    };

    if (combinedProps.itemKey == null) {
      combinedProps.itemKey = element.key == null ? index : element.key;
    }

    return (
      <Item
        key={combinedProps.itemKey}
        isFirst={!Array.isArray(children) || index === 0}
        isLast={!Array.isArray(children) || index === children.length - 1}
        {...combinedProps}
      />
    );
  };

  return {
    expanded: containerProps.activeKey != null,
    children: Array.isArray(children) ? children.map(toItem) : toItem(children),
  };
};

const onActivate = ({ activateKey }) => (event, value) => {
  if (event) {
    event.stopPropagation();
  }
  activateKey(value);
};

export default compose(
  withState('activeKey', 'activateKey', ({ defaultActiveKey }) => defaultActiveKey),
  withHandlers({ onActivate }),
  embedHandler('onActivate', 'onChange'),
  withProps(propsMapper),
  withProps(stylesMapper),
);
