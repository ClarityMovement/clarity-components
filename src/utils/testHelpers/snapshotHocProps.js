import React from 'react';

import enzymeToJson from 'enzyme-to-json';
import { shallowWithUntil as shallow } from './until';

const snapshotPropsMapperHoc = (hoc, inputProps) => {
  // This Component can't render deeply, must work with shallow renderer.
  const PropsDisplayer = outputProps => (
    <div name="props-displayer">
      <input-props {...inputProps} />
      <output-props {...outputProps} />
    </div>
  );
  const NewComponent = hoc(PropsDisplayer);
  const wrapper = shallow(<NewComponent {...inputProps} />).until(PropsDisplayer);
  expect(enzymeToJson(wrapper)).toMatchSnapshot();
};

export default snapshotPropsMapperHoc;
