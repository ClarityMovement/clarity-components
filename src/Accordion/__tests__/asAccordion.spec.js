/* eslint-env jest */
import React from 'react';
import snapshotHocProps from 'react-render-counter/utils/testHelpers/snapshotHocProps';
import asAccordion from '../asAccordion';

describe('asAccordion(BaseComponent): NewComponent', () => {
  let props;
  let children;
  beforeEach(() => {
    children = [
      <div key="1" />,
      <div key="2" />,
    ];
  });

  afterEach(() => snapshotHocProps(asAccordion, props));

  describe('with narrow screen', () => {
    test('with default props', () => {
      props = { children };
    });

    test('set default activeKey', () => {
      props = {
        defaultActiveKey: 1,
        children,
      };
    });
  });

  describe('with wide screen', () => {
    beforeEach(() => {
      props.wide = true;
    });

    test('with default props', () => {
      props = {
        ...props,
        children,
      };
    });
  });
});
