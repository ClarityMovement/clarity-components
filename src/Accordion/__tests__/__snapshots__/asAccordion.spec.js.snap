// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`asAccordion(BaseComponent): NewComponent with narrow screen set default activeKey 1`] = `
<div
  name="props-displayer"
>
  <input-props
    defaultActiveKey="1"
  >
    <div />
    <div />
    <div
      itemKey={2}
    />
  </input-props>
  <output-props
    activateKey={[Function]}
    activeKey="1"
    defaultActiveKey="1"
    expanded={true}
    onActivate={[Function]}
    style={
      Object {
        "alignItems": "center",
        "bottom": 0,
        "display": "flex",
        "flexDirection": "column",
        "height": "100%",
        "padding": 0,
        "position": "fixed",
        "top": 0,
        "width": "100%",
      }
    }
  >
    <withProps(onlyUpdateForKeys(withHandlers(withHandlers(withProps(div)))))
      activateKey={[Function]}
      activeKey="1"
      defaultActiveKey="1"
      isFirst={true}
      isLast={false}
      itemKey="0"
      onActivate={[Function]}
    />
    <withProps(onlyUpdateForKeys(withHandlers(withHandlers(withProps(div)))))
      activateKey={[Function]}
      activeKey="1"
      defaultActiveKey="1"
      isFirst={false}
      isLast={false}
      itemKey={1}
      onActivate={[Function]}
    />
    <withProps(onlyUpdateForKeys(withHandlers(withHandlers(withProps(div)))))
      activateKey={[Function]}
      activeKey="1"
      defaultActiveKey="1"
      isFirst={false}
      isLast={true}
      itemKey={2}
      onActivate={[Function]}
    />
  </output-props>
</div>
`;

exports[`asAccordion(BaseComponent): NewComponent with narrow screen with default props and single child 1`] = `
<div
  name="props-displayer"
>
  <input-props>
    <div />
  </input-props>
  <output-props
    activateKey={[Function]}
    expanded={false}
    onActivate={[Function]}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "column",
        "padding": 0,
        "width": "100%",
      }
    }
  >
    <withProps(onlyUpdateForKeys(withHandlers(withHandlers(withProps(div)))))
      activateKey={[Function]}
      isFirst={true}
      isLast={true}
      onActivate={[Function]}
    />
  </output-props>
</div>
`;

exports[`asAccordion(BaseComponent): NewComponent with wide screen set default activeKey 1`] = `
<div
  name="props-displayer"
>
  <input-props
    defaultActiveKey={1}
    wide={true}
  >
    <div />
    <div />
    <div
      itemKey={2}
    />
  </input-props>
  <output-props
    activateKey={[Function]}
    activeKey={1}
    defaultActiveKey={1}
    expanded={true}
    onActivate={[Function]}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "column",
        "height": "100%",
        "margin": "0 auto",
        "padding": 0,
        "width": "95%",
      }
    }
    wide={true}
  >
    <withProps(onlyUpdateForKeys(withHandlers(withHandlers(withProps(div)))))
      activateKey={[Function]}
      activeKey={1}
      defaultActiveKey={1}
      isFirst={true}
      isLast={false}
      itemKey="0"
      onActivate={[Function]}
      wide={true}
    />
    <withProps(onlyUpdateForKeys(withHandlers(withHandlers(withProps(div)))))
      activateKey={[Function]}
      activeKey={1}
      defaultActiveKey={1}
      isFirst={false}
      isLast={false}
      itemKey={1}
      onActivate={[Function]}
      wide={true}
    />
    <withProps(onlyUpdateForKeys(withHandlers(withHandlers(withProps(div)))))
      activateKey={[Function]}
      activeKey={1}
      defaultActiveKey={1}
      isFirst={false}
      isLast={true}
      itemKey={2}
      onActivate={[Function]}
      wide={true}
    />
  </output-props>
</div>
`;

exports[`asAccordion(BaseComponent): NewComponent with wide screen with default props and single child 1`] = `
<div
  name="props-displayer"
>
  <input-props
    wide={true}
  >
    <div />
  </input-props>
  <output-props
    activateKey={[Function]}
    expanded={false}
    onActivate={[Function]}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "column",
        "margin": "0 auto",
        "padding": 0,
        "width": "95%",
      }
    }
    wide={true}
  >
    <withProps(onlyUpdateForKeys(withHandlers(withHandlers(withProps(div)))))
      activateKey={[Function]}
      isFirst={true}
      isLast={true}
      onActivate={[Function]}
      wide={true}
    />
  </output-props>
</div>
`;
